File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.go
func evaluate(s string, knowledge [][]string) string {
    mp := make(map[string]string)

    for i:=0;i<len(knowledge);i++{
        mp[knowledge[i][0]] = knowledge[i][1]
    }

    ans := ""
    n := len(s)
    for i:=0;i<n;i++{
        if s[i] == '(' {
            var start int = i+1
            for i<n && s[i]!=')'{
                i++
            }

            var end int = i
            var key string = s[start : end];

            if value,ok := mp[key];ok{
                ans += value
            }else{
                ans += "?"
            }
        }else{
            ans += string(s[i])
        }
    }

    return ans
}

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.cpp
class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        unordered_map<string , string> mp;
        for(auto& val : knowledge){
            mp[val[0]] = val[1];
        }

        string ans = "";
        for(int i=0;i<s.size();i++){
            if(s[i]>='a' && s[i]<='z'){
                ans.push_back(s[i]);
                continue;
            }

            int braceStart = i+1;
            while(i<s.size() && s[i] != ')'){
                i++;
            }
            int braceEnd = i-1;
            string key = s.substr(braceStart , braceEnd-braceStart+1);
            if(mp.count(key)){
                ans += mp[key];
            }
            else{
                ans.push_back('?');
            }
            
        }

        return ans;
    }
};